File: csharp/tests/Idkollen.Client.Tests/MitIdTests.cs
using System.Net;
using Idkollen.Client.Models;
using Xunit;

namespace Idkollen.Client.Tests;

public class MitIdTests
{
    private static (Idkollen.Client.IdkollenClient Client, MockHttpMessageHandler Handler) Build()
    {
        var handler = new MockHttpMessageHandler();
        var http = new HttpClient(handler);
        var client = new Idkollen.Client.IdkollenClientBuilder("cid", "sec")
            .BaseUrl("https://x.test").HttpClient(http).Build();
        return (client, handler);
    }

    [Fact]
    public async Task AuthPending()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new { status = "PENDING", id = "m1", url = "https://login" });
        var result = await client.MitId.AuthAsync(new MitIdAuthRequest());
        Assert.IsType<MitIdPending>(result);
    }

    [Fact]
    public async Task SignCompletedWithSignResult()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new
        {
            status = "COMPLETED", id = "m1", ssn = "0101010000",
            name = "Hans", givenName = "Hans", surname = "H",
            signResult = new { checksum = "abc123" },
        });
        var result = await client.MitId.SignStatusAsync("m1");
        var done = Assert.IsType<MitIdCompleted>(result);
        Assert.NotNull(done.SignResult);
        Assert.Equal("abc123", done.SignResult!.Checksum);
    }
}

File: swift/Tests/IdkollenClientTests/MitIdTests.swift
import XCTest
@testable import IdkollenClient

final class MitIdTests: XCTestCase {
    var client: IdkollenClient!

    override func setUp() {
        super.setUp()
        MockURLProtocol.reset()
        client = IdkollenClientBuilder(clientId: "cid", clientSecret: "sec")
            .baseUrl("https://x.test")
            .urlSession(.mocked())
            .build()
    }

    func testAuthPending() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "status": "PENDING",
            "id": "abc",
            "url": "https://login",
            "bindingMessage": "approve this",
        ])

        let result = try await client.mitId.auth(MitIdAuthRequest())
        guard case .pending(let p) = result else { return XCTFail("expected pending") }
        XCTAssertEqual(p.url, "https://login")
    }

    func testSignCompletedWithSignResult() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "status": "COMPLETED",
            "id": "abc",
            "ssn": "1234567890",
            "name": "Test User",
            "givenName": "Test",
            "surname": "User",
            "signResult": ["checksum": "abc123"],
        ])

        let result = try await client.mitId.sign(MitIdSignRequest(text: "please sign"))
        if case .completed(let done) = result {
            XCTAssertEqual(done.signResult?.checksum, "abc123")
        } else {
            XCTFail("expected completed")
        }
    }
}
